CMakeLists.txt

```diff
@@ -1,93 +1,41 @@
-######### CMake Version #####################
-cmake_minimum_required(VERSION 2.8.11)
-#############################################
+project(eigen-cuda LANGUAGES CXX)
+cmake_minimum_required(VERSION 3.14)
 
-######### Options ###########################
-option( CORE_USE_CUDA    "Use CUDA to speed up certain parts of the code."              ON )
-#############################################
+option(USE_CUDA "Whether to use CUDA" OFF)
 
-
-######### CUDA decisions ####################
-if (CORE_USE_CUDA)
-    MESSAGE( STATUS ">> --------------  USING CUDA --------------" )
-    set( CUDA_TOOLKIT_ROOT_DIR "/opt/cuda" )
-    if (APPLE OR UNIX)
-        set(CMAKE_C_COMPILER   /opt/cuda/bin/gcc)
-        set(CMAKE_CXX_COMPILER /opt/cuda/bin/g++)
-    elseif (WIN32)
-        ### By default we use VS
-        MESSAGE( STATUS ">> User compiler:             MSVC" )
-        MESSAGE( STATUS ">> Choosing a different compiler is not yet implemented for Windows" )
-    endif()
-endif()
-#############################################
-
-
-######### Info ##############################
-MESSAGE( STATUS ">> CMAKE_C_COMPILER:        " ${CMAKE_C_COMPILER} )
-MESSAGE( STATUS ">> CMAKE_CXX_COMPILER:      " ${CMAKE_CXX_COMPILER} )
-#############################################
-
-
-######### Project Name ######################
-project(eigencuda)
-SET( EXECUTABLE_NAME run )
-#############################################
-
-
-### Find includes in corresponding build directories
-set( CMAKE_INCLUDE_CURRENT_DIR ON )
-######### Have the binary placed into the source head
-set( CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_SOURCE_DIR} )
-### Output paths for multi-config builds (e.g. msvc)
-foreach( OUTPUTCONFIG ${CMAKE_CONFIGURATION_TYPES} )
-    string( TOUPPER ${OUTPUTCONFIG} OUTPUTCONFIG )
-    set( CMAKE_RUNTIME_OUTPUT_DIRECTORY_${OUTPUTCONFIG} ${CMAKE_SOURCE_DIR} )
-    # set( CMAKE_LIBRARY_OUTPUT_DIRECTORY_${OUTPUTCONFIG} ${youroutputdirectory} )
-    # set( CMAKE_ARCHIVE_OUTPUT_DIRECTORY_${OUTPUTCONFIG} ${youroutputdirectory} )
-endforeach( OUTPUTCONFIG CMAKE_CONFIGURATION_TYPES )
-#############################################
-set( CMAKE_DISABLE_SOURCE_CHANGES  ON )
-set( CMAKE_DISABLE_IN_SOURCE_BUILD ON )
-#############################################
-
-
-######### CUDA decisions ####################
-if (CORE_USE_CUDA)
-    find_package(CUDA REQUIRED)
+if (USE_CUDA)
+    enable_language(CUDA)
     add_definitions(-DUSE_CUDA)
 endif()
-#############################################
 
-#############################################
-set(SOURCE_FILES
-    src/test.cpp
-    src/kernel.cu
-    src/kernel.cpp
+include(FetchContent)
+FetchContent_Declare(
+    Eigen3
+    GIT_REPOSITORY git@gitlab.com:libeigen/eigen.git
+    GIT_TAG master
```

Comment:
I'm not sure about fetching a branch instead of a commit or tag. It will mean people will get varying results over time.

You say
> couldn't build it due to changes in Eigen/CMake

But this would change the Eigen version for people over time, without any relation to the CUDA version -- assuming an old CUDA version I believe this example should still build, even on newer CMake. Newer CUDA versions will require newer Eigen versions, but unfortunately it is also my experience that newer Eigen versions break on older CUDA.
Maybe it would be better to specify with which CUDA versions a given Eigen version should work?

Unfortunately, I don't see a way to future-proof this except maybe using conan or vcpkg, which seems a bit overkill...

Reply:
is there a table that maps cuda / eigen versions somehow?

Reply:
I'm not aware of one and I think it would be overkill to implement such a mapping here.
For me it would be fine to choose a version and write a short comment with the CUDA version with which this MWE works.